File: src/lib/bungie/syncInventory.ts
import { computeRollTags } from "@/lib/inventory/rollTags";
import { buildStoredSocketPlugs } from "@/lib/inventory/instances/buildStoredSocketPlugs";
import { loadWeaponSocketContext } from "@/lib/inventory/instances/weaponSocketContext";
import type { StoredSocketPlug } from "@/lib/inventory/instances/types";
import { isWeaponBucketHash } from "@/lib/bungie/inventoryBuckets";
import type { AppDatabase } from "@/lib/db/client";
import { upsertInventoryBatch, getInventoryStatus } from "@/lib/db/repositories/inventoryRepository";
import { updateUserMembership } from "@/lib/db/repositories/userRepository";
import type { DbUser, UserInventoryItem } from "@/lib/db/types";
import {
  inventoryBucketLabel,
  SUBCLASS_BUCKET_HASH,
} from "@/lib/bungie/inventoryBuckets";
import {
  buildEquipmentBucketLookup,
  needsEquipmentBucketResolution,
  resolveTransferContainerBuckets,
} from "@/lib/bungie/resolveEquipmentBuckets";
import type {
  BungieProfileClient,
  DestinyMembership,
  InventoryParseDiagnostics,
  RawInventoryItem,
} from "@/lib/bungie/types";
import type { EntityCache, ManifestService } from "@/lib/manifest/types/services";
import type { ExoticWeaponRecord, WeaponRecord } from "@/lib/manifest/types/records";

type WeaponCatalogRecord = WeaponRecord | ExoticWeaponRecord;

const syncLocks = new Map<number, Promise<SyncInventoryResult>>();

export interface SyncInventoryResult {
  itemCount: number;
  syncVersion: number;
  lastFullSyncAt: string;
  diagnostics: InventoryParseDiagnostics;
}

export class SyncInProgressError extends Error {
  constructor() {
    super("Inventory sync already in progress for this user");
    this.name = "SyncInProgressError";
  }
}


function bucketLabel(bucketHash: number): string {
  return inventoryBucketLabel(bucketHash);
}

async function buildWeaponLookup(cache: EntityCache): Promise<Map<number, WeaponCatalogRecord>> {
  const [weapons, exoticWeapons] = await Promise.all([
    cache.getStore("weapons"),
    cache.getStore("exotic-weapons"),
  ]);
  const lookup = new Map<number, WeaponCatalogRecord>();
  for (const weapon of [...weapons, ...exoticWeapons]) {
    lookup.set(weapon.hash, weapon);
  }
  return lookup;
}

async function buildExoticArmorHashSet(cache: EntityCache): Promise<Set<number>> {
  const armor = await cache.getStore("exotic-armor");
  return new Set(armor.map((piece) => piece.hash));
}

function enrichManifestDiagnostics(
  diagnostics: InventoryParseDiagnostics,
  rawItems: RawInventoryItem[],
  weaponLookup: Map<number, WeaponCatalogRecord>,
  exoticArmorHashes: Set<number>,
): void {
  const equipmentHashes = new Set<number>();
  for (const item of rawItems) {
    if (item.bucketHash === SUBCLASS_BUCKET_HASH) continue;
    equipmentHashes.add(item.itemHash);
  }

  let inWeaponsCatalog = 0;
  let inExoticArmorCatalog = 0;
  for (const hash of equipmentHashes) {
    if (weaponLookup.has(hash)) {
      inWeaponsCatalog += 1;
    }
    if (exoticArmorHashes.has(hash)) {
      inExoticArmorCatalog += 1;
    }
  }

  const matchedHashes = new Set<number>();
  for (const hash of equipmentHashes) {
    if (weaponLookup.has(hash) || exoticArmorHashes.has(hash)) {
      matchedHashes.add(hash);
    }
  }

  diagnostics.manifest = {
    equipmentItemHashes: equipmentHashes.size,
    inWeaponsCatalog,
    inExoticArmorCatalog,
    unmatchedEquipmentHashes: equipmentHashes.size - matchedHashes.size,
  };
}

function logInventorySyncDiagnostics(diagnostics: InventoryParseDiagnostics): void {
  const topUnknownBuckets = Object.entries(diagnostics.dropped.unknownBuckets)
    .sort((a, b) => b[1] - a[1])
    .slice(0, 15);
  console.info(
    "[inventory-sync]",
    JSON.stringify(
      {
        membership: diagnostics.membership,
        rawTotal: diagnostics.raw.total,
        parsedTotal: diagnostics.parsed.total,
        equipmentTotal: diagnostics.parsed.equipmentTotal,
        subclassTotal: diagnostics.parsed.subclassTotal,
        droppedTotal: diagnostics.dropped.total,
        dropped: {
          invalidShape: diagnostics.dropped.invalidShape,
          unknownBucket: diagnostics.dropped.unknownBucket,
          missingInstanceId: diagnostics.dropped.missingInstanceId,
          topUnknownBuckets,
        },
        byLocation: diagnostics.parsed.byLocation,
        byBucket: diagnostics.parsed.byBucket,
        resolution: diagnostics.resolution,
        manifest: diagnostics.manifest,
      },
      null,
      2,
    ),
  );
}

async function buildPerkNameMap(cache: EntityCache): Promise<Map<number, string>> {
  const perks = await cache.getStore("weapon-perks");
  return new Map(perks.map((p) => [p.hash, p.name]));
}

function normalizeItems(
  rawItems: RawInventoryItem[],
  perkNameMap: Map<number, string>,
  weaponLookup: Map<number, WeaponCatalogRecord>,
  syncedAt: string,
  socketPlugsByInstance: Map<string, StoredSocketPlug[] | null>,
): UserInventoryItem[] {
  return rawItems.map((raw) => {
    const catalogEntry = weaponLookup.get(raw.itemHash) ?? null;
    // Legendary WeaponRecord only — exotics also have perkColumns but not roll columns.
    const weaponRecord =
      catalogEntry && "originTraitHashes" in catalogEntry
        ? (catalogEntry as WeaponRecord)
        : null;
    const rollTags = computeRollTags(raw.plugHashes, perkNameMap, weaponRecord, {
      isCrafted: raw.isCrafted,
    });
    return {
      instanceId: raw.instanceId,
      itemHash: raw.itemHash,
      bucket: bucketLabel(raw.bucketHash),
      location: raw.location,
      characterId: raw.characterId,
      power: raw.power,
      isMasterwork: raw.isMasterwork,
      isCrafted: raw.isCrafted,
      plugHashes: raw.plugHashes,
      rollTags,
      statValues: raw.statValues,
      gearTier: raw.gearTier ?? null,
      socketPlugs: socketPlugsByInstance.get(raw.instanceId) ?? null,
      syncedAt,
    };
  });
}

async function buildSocketPlugsForItems(
  rawItems: RawInventoryItem[],
  manifest: ManifestService,
  manifestVersion: string,
): Promise<Map<string, StoredSocketPlug[] | null>> {
  const byInstance = new Map<string, StoredSocketPlug[] | null>();
  const contextCache = new Map<
    number,
    Awaited<ReturnType<typeof loadWeaponSocketContext>>
  >();

  for (const raw of rawItems) {
    if (!isWeaponBucketHash(raw.bucketHash) || !raw.socketCapture?.length) {
      byInstance.set(raw.instanceId, null);
      continue;
    }

    const allPlugHashes = [
      ...new Set(
        raw.socketCapture.flatMap((row) => [row.equippedPlugHash, ...row.reusablePlugHashes]),
      ),
    ];

    let ctx = contextCache.get(raw.itemHash);
    if (!ctx) {
      ctx = await loadWeaponSocketContext(manifest, manifestVersion, raw.itemHash, allPlugHashes);
      contextCache.set(raw.itemHash, ctx);
    } else {
      const missing = allPlugHashes.filter((hash) => !ctx!.plugCategoryByHash.has(hash));
      if (missing.length > 0) {
        const extra = await loadWeaponSocketContext(
          manifest,
          manifestVersion,
          raw.itemHash,
          [...allPlugHashes, ...missing],
        );
        for (const [hash, category] of extra.plugCategoryByHash) {
          ctx.plugCategoryByHash.set(hash, category);
        }
      }
    }

    const stored = buildStoredSocketPlugs({
      socketCapture: raw.socketCapture,
      plugCategoryByHash: ctx.plugCategoryByHash,
      plugItemTypeByHash: ctx.plugItemTypeByHash,
      weaponPerkSocketIndexes: ctx.weaponPerkSocketIndexes,
    });
    byInstance.set(raw.instanceId, stored);
  }

  return byInstance;
}

async function resolveDestinyMembership(
  profileClient: BungieProfileClient,
  accessToken: string,
): Promise<DestinyMembership> {
  const memberships = await profileClient.getMemberships(accessToken);
  const membership = memberships[0];
  if (!membership) {
    throw new Error("No Destiny memberships found");
  }
  return membership;
}

async function performSync(
  db: AppDatabase,
  user: DbUser,
  accessToken: string,
  profileClient: BungieProfileClient,
  entityCache: EntityCache,
  manifest: ManifestService,
  manifestVersion: string,
): Promise<SyncInventoryResult> {
  const membership = await resolveDestinyMembership(profileClient, accessToken);

  if (
    user.membershipType !== membership.membershipType ||
    user.displayName !== membership.displayName
  ) {
    updateUserMembership(db, user.id, membership.membershipType, membership.displayName);
  }

  const { items: rawItems, diagnostics } = await profileClient.getFullInventoryWithDiagnostics(
    accessToken,
    membership,
  );

  const transferItemHashes = rawItems
    .filter((item) => needsEquipmentBucketResolution(item.bucketHash))
    .map((item) => item.itemHash);
  const equipmentBucketLookup = await buildEquipmentBucketLookup(
    manifest,
    manifestVersion,
    transferItemHashes,
  );
  const resolved = resolveTransferContainerBuckets(rawItems, equipmentBucketLookup);

  const [perkNameMap, weaponLookup, exoticArmorHashes] = await Promise.all([
    buildPerkNameMap(entityCache),
    buildWeaponLookup(entityCache),
    buildExoticArmorHashSet(entityCache),
  ]);
  enrichManifestDiagnostics(diagnostics, resolved.items, weaponLookup, exoticArmorHashes);

  const syncedAt = new Date().toISOString();
  const socketPlugsByInstance = await buildSocketPlugsForItems(
    resolved.items,
    manifest,
    manifestVersion,
  );
  const items = normalizeItems(
    resolved.items,
    perkNameMap,
    weaponLookup,
    syncedAt,
    socketPlugsByInstance,
  );
  diagnostics.resolution = {
    resolvedFromTransfer: resolved.resolvedFromTransfer,
    droppedNonEquipment: resolved.droppedNonEquipment,
    storedTotal: items.length,
    storedEquipment: items.filter((item) => item.bucket !== "Subclass").length,
  };
  logInventorySyncDiagnostics(diagnostics);
  upsertInventoryBatch(db, user.id, items);

  const status = getInventoryStatus(db, user.id);
  return {
    itemCount: status?.itemCount ?? items.length,
    syncVersion: status?.syncVersion ?? 1,
    lastFullSyncAt: status?.lastFullSyncAt ?? syncedAt,
    diagnostics,
  };
}

export async function syncUserInventory(
  db: AppDatabase,
  user: DbUser,
  accessToken: string,
  profileClient: BungieProfileClient,
  entityCache: EntityCache,
  manifest: ManifestService,
  manifestVersion: string,
): Promise<SyncInventoryResult> {
  const existing = syncLocks.get(user.id);
  if (existing) {
    throw new SyncInProgressError();
  }

  const syncPromise = performSync(
    db,
    user,
    accessToken,
    profileClient,
    entityCache,
    manifest,
    manifestVersion,
  ).finally(() => {
    syncLocks.delete(user.id);
  });

  syncLocks.set(user.id, syncPromise);
  return syncPromise;
}

/** Test helper: clear in-memory sync mutex state. */
export function clearSyncLocksForTests(): void {
  syncLocks.clear();
}

File: packages/bungie/lib/src/sync/sync_inventory.dart
import 'dart:async';

import 'package:destiny2_db/destiny2_db.dart';

import '../inventory/build_stored_socket_plugs.dart';
import '../inventory/roll_tag_lookups.dart';
import '../inventory/roll_tags.dart';
import '../inventory/weapon_socket_context.dart';
import '../profile/bungie_profile_client.dart';
import '../profile/equipment_bucket_lookup.dart';
import '../profile/inventory_buckets.dart';
import '../profile/inventory_parse.dart';
import '../profile/profile_types.dart';

/// Result of a successful full inventory sync into Drift.
class SyncInventoryResult {
  const SyncInventoryResult({
    required this.itemCount,
    required this.syncVersion,
    required this.lastFullSyncAt,
    required this.diagnostics,
  });

  final int itemCount;
  final int syncVersion;
  final String lastFullSyncAt;
  final InventoryParseDiagnostics diagnostics;
}

/// Thrown when inventory sync is already in progress for the user.
///
/// Mirrors product `SyncInProgressError` / DART-016 busy lock.
class SyncInProgressError implements Exception {
  SyncInProgressError([this.message = 'Inventory sync already in progress for this user']);

  final String message;

  @override
  String toString() => 'SyncInProgressError: $message';
}

/// Resolve Destiny membership, fetch inventory, full-replace into Drift.
///
/// - Uses first membership from [BungieProfileClient.getMemberships] (primary-sorted).
/// - Updates local user membership type/display when changed.
/// - [equipmentBucketLookup] / [equipmentBucketLookupBuilder] map itemHash → equipment
///   bucketHash for vault/postmaster resolution.
/// - **Production hosts MUST wire a non-empty lookup** whenever entity/manifest data
///   is available (DART-050 / GAP-INV-01). Empty lookup is test-only — transfer
///   containers (vault General / postmaster) are dropped before Drift write.
/// - [perkNameMap] / [perkNameMapBuilder] + [weaponRollMetaLookup] /
///   [weaponRollMetaLookupBuilder] supply DART-051 roll tag enrichment inputs
///   (Next `computeRollTags` parity). Empty maps → Crafted-only when `isCrafted`.
/// - [weaponSocketContextBuilder] supplies DART-052 socket plug enrichment
///   (Next `buildStoredSocketPlugs` + `loadWeaponSocketContext`). Without it,
///   weapons keep raw capture maps (no columnKind) — incomplete for perk grids.
/// - [now] ISO-8601 timestamp written as syncedAt / lastFullSyncAt (injectable for tests).
Future<SyncInventoryResult> syncUserInventory({
  required AppDatabase db,
  required int userId,
  required String accessToken,
  required BungieProfileClient profileClient,
  Map<int, int>? equipmentBucketLookup,
  EquipmentBucketLookupBuilder? equipmentBucketLookupBuilder,
  Map<int, String>? perkNameMap,
  PerkNameMapBuilder? perkNameMapBuilder,
  Map<int, RollTagWeaponMeta>? weaponRollMetaLookup,
  WeaponRollMetaLookupBuilder? weaponRollMetaLookupBuilder,
  WeaponSocketContextBuilder? weaponSocketContextBuilder,
  String? now,
  InventoryBusyLock? lock,
}) async {
  try {
    return await (lock ?? defaultInventoryBusyLock).runExclusive(userId, () async {
      return _performSync(
        db: db,
        userId: userId,
        accessToken: accessToken,
        profileClient: profileClient,
        equipmentBucketLookup: equipmentBucketLookup,
        equipmentBucketLookupBuilder: equipmentBucketLookupBuilder,
        perkNameMap: perkNameMap,
        perkNameMapBuilder: perkNameMapBuilder,
        weaponRollMetaLookup: weaponRollMetaLookup,
        weaponRollMetaLookupBuilder: weaponRollMetaLookupBuilder,
        weaponSocketContextBuilder: weaponSocketContextBuilder,
        now: now ?? DateTime.now().toUtc().toIso8601String(),
      );
    });
  } on InventoryReplaceBusyException {
    throw SyncInProgressError();
  }
}

Future<SyncInventoryResult> _performSync({
  required AppDatabase db,
  required int userId,
  required String accessToken,
  required BungieProfileClient profileClient,
  Map<int, int>? equipmentBucketLookup,
  EquipmentBucketLookupBuilder? equipmentBucketLookupBuilder,
  Map<int, String>? perkNameMap,
  PerkNameMapBuilder? perkNameMapBuilder,
  Map<int, RollTagWeaponMeta>? weaponRollMetaLookup,
  WeaponRollMetaLookupBuilder? weaponRollMetaLookupBuilder,
  WeaponSocketContextBuilder? weaponSocketContextBuilder,
  required String now,
}) async {
  final memberships = await profileClient.getMemberships(accessToken);
  if (memberships.isEmpty) {
    throw StateError('No Destiny memberships found');
  }
  final membership = memberships.first;

  final user = await getUser(db, userId);
  if (user == null) {
    throw StateError('User $userId not found');
  }
  if (user.membershipType != membership.membershipType ||
      user.displayName != membership.displayName) {
    await updateUserMembership(
      db,
      userId,
      membershipType: membership.membershipType,
      displayName: membership.displayName,
    );
  }

  final parsed = await profileClient.getFullInventoryWithDiagnostics(
    accessToken,
    membership,
  );

  final transferHashes = <int>[
    for (final item in parsed.items)
      if (needsEquipmentBucketResolution(item.bucketHash)) item.itemHash,
  ];
  final lookup = await _resolveEquipmentBucketLookup(
    transferHashes: transferHashes,
    equipmentBucketLookup: equipmentBucketLookup,
    equipmentBucketLookupBuilder: equipmentBucketLookupBuilder,
  );

  final resolved = resolveTransferContainerBuckets(
    parsed.items,
    lookup,
  );

  final plugHashes = <int>[
    for (final item in resolved.items) ...item.plugHashes,
  ];
  final itemHashes = <int>[
    for (final item in resolved.items) item.itemHash,
  ];
  final resolvedPerkMap = await _resolvePerkNameMap(
    plugHashes: plugHashes,
    perkNameMap: perkNameMap,
    perkNameMapBuilder: perkNameMapBuilder,
  );
  final resolvedWeaponMeta = await _resolveWeaponRollMetaLookup(
    itemHashes: itemHashes,
    weaponRollMetaLookup: weaponRollMetaLookup,
    weaponRollMetaLookupBuilder: weaponRollMetaLookupBuilder,
  );

  final socketPlugsByInstance = await _buildSocketPlugsForItems(
    resolved.items,
    weaponSocketContextBuilder: weaponSocketContextBuilder,
  );

  final records = _normalizeItems(
    resolved.items,
    now,
    perkNameMap: resolvedPerkMap,
    weaponRollMetaLookup: resolvedWeaponMeta,
    socketPlugsByInstance: socketPlugsByInstance,
  );

  parsed.diagnostics.resolution = InventoryResolutionCounts(
    resolvedFromTransfer: resolved.resolvedFromTransfer,
    droppedNonEquipment: resolved.droppedNonEquipment,
    storedTotal: records.length,
    storedEquipment:
        records.where((i) => i.bucket != inventoryBucketLabel(kSubclassBucketHash)).length,
  );

  // Exclusive lock already held — use non-exclusive replace inside.
  final status = await replaceInventoryBatch(
    db,
    userId,
    items: records,
    now: now,
  );

  return SyncInventoryResult(
    itemCount: status.itemCount,
    syncVersion: status.syncVersion,
    lastFullSyncAt: status.lastFullSyncAt ?? now,
    diagnostics: parsed.diagnostics,
  );
}

/// Merge explicit map with async builder (explicit entries win on conflict).
Future<Map<int, int>> _resolveEquipmentBucketLookup({
  required List<int> transferHashes,
  Map<int, int>? equipmentBucketLookup,
  EquipmentBucketLookupBuilder? equipmentBucketLookupBuilder,
}) async {
  final merged = <int, int>{};
  if (equipmentBucketLookupBuilder != null && transferHashes.isNotEmpty) {
    merged.addAll(await equipmentBucketLookupBuilder(transferHashes));
  }
  if (equipmentBucketLookup != null) {
    merged.addAll(equipmentBucketLookup);
  }
  return merged;
}

Future<Map<int, String>> _resolvePerkNameMap({
  required List<int> plugHashes,
  Map<int, String>? perkNameMap,
  PerkNameMapBuilder? perkNameMapBuilder,
}) async {
  final merged = <int, String>{};
  if (perkNameMapBuilder != null && plugHashes.isNotEmpty) {
    merged.addAll(await perkNameMapBuilder(plugHashes));
  }
  if (perkNameMap != null) {
    merged.addAll(perkNameMap);
  }
  return merged;
}

Future<Map<int, RollTagWeaponMeta>> _resolveWeaponRollMetaLookup({
  required List<int> itemHashes,
  Map<int, RollTagWeaponMeta>? weaponRollMetaLookup,
  WeaponRollMetaLookupBuilder? weaponRollMetaLookupBuilder,
}) async {
  final merged = <int, RollTagWeaponMeta>{};
  if (weaponRollMetaLookupBuilder != null && itemHashes.isNotEmpty) {
    merged.addAll(await weaponRollMetaLookupBuilder(itemHashes));
  }
  if (weaponRollMetaLookup != null) {
    merged.addAll(weaponRollMetaLookup);
  }
  return merged;
}

/// Next parity: `buildSocketPlugsForItems` — weapons only; null for non-weapons.
Future<Map<String, List<Map<String, Object?>>?>> _buildSocketPlugsForItems(
  List<RawInventoryItem> rawItems, {
  WeaponSocketContextBuilder? weaponSocketContextBuilder,
}) async {
  final byInstance = <String, List<Map<String, Object?>>?>{};
  final contextCache = <int, WeaponSocketContext>{};

  for (final raw in rawItems) {
    if (!isWeaponBucketHash(raw.bucketHash) ||
        raw.socketCapture == null ||
        raw.socketCapture!.isEmpty) {
      byInstance[raw.instanceId] = null;
      continue;
    }

    final capture = raw.socketCapture!;
    final allPlugHashes = <int>{
      for (final row in capture) ...[
        row.equippedPlugHash,
        ...row.reusablePlugHashes,
      ],
    }.toList(growable: false);

    if (weaponSocketContextBuilder == null) {
      // Degradation: raw capture without columnKind (incomplete for perk grids).
      byInstance[raw.instanceId] =
          capture.map((s) => s.toJsonMap()).toList(growable: false);
      continue;
    }

    var ctx = contextCache[raw.itemHash];
    if (ctx == null) {
      ctx = await weaponSocketContextBuilder(raw.itemHash, allPlugHashes);
      contextCache[raw.itemHash] = ctx;
    } else {
      final missing = allPlugHashes
          .where((h) => !ctx!.plugCategoryByHash.containsKey(h))
          .toList(growable: false);
      if (missing.isNotEmpty) {
        final extra = await weaponSocketContextBuilder(
          raw.itemHash,
          [...allPlugHashes, ...missing],
        );
        ctx = ctx.mergePlugMaps(extra);
        contextCache[raw.itemHash] = ctx;
      }
    }

    final hasEnrichment = ctx.plugCategoryByHash.isNotEmpty ||
        ctx.weaponPerkSocketIndexes.isNotEmpty;
    if (!hasEnrichment) {
      byInstance[raw.instanceId] =
          capture.map((s) => s.toJsonMap()).toList(growable: false);
      continue;
    }

    final stored = buildStoredSocketPlugs(
      socketCapture: capture,
      plugCategoryByHash: ctx.plugCategoryByHash,
      plugItemTypeByHash: ctx.plugItemTypeByHash,
      weaponPerkSocketIndexes: ctx.weaponPerkSocketIndexes,
    );
    byInstance[raw.instanceId] =
        stored.map((p) => p.toJsonMap()).toList(growable: false);
  }

  return byInstance;
}

List<InventoryItemRecord> _normalizeItems(
  List<RawInventoryItem> rawItems,
  String syncedAt, {
  Map<int, String> perkNameMap = const {},
  Map<int, RollTagWeaponMeta> weaponRollMetaLookup = const {},
  Map<String, List<Map<String, Object?>>?> socketPlugsByInstance = const {},
}) {
  return rawItems.map((raw) {
    final rollTags = computeRollTags(
      raw.plugHashes,
      perkNameMap,
      weapon: weaponRollMetaLookup[raw.itemHash],
      isCrafted: raw.isCrafted,
    );
    final socketPlugs = socketPlugsByInstance.containsKey(raw.instanceId)
        ? socketPlugsByInstance[raw.instanceId]
        : (!isWeaponBucketHash(raw.bucketHash)
            ? null
            : raw.socketCapture
                ?.map((s) => s.toJsonMap())
                .toList(growable: false));
    return InventoryItemRecord(
      instanceId: raw.instanceId,
      itemHash: raw.itemHash,
      bucket: inventoryBucketLabel(raw.bucketHash),
      location: raw.location,
      characterId: raw.characterId,
      power: raw.power,
      isMasterwork: raw.isMasterwork,
      isCrafted: raw.isCrafted,
      plugHashes: raw.plugHashes,
      rollTags: rollTags,
      statValues: raw.statValues,
      gearTier: raw.gearTier,
      socketPlugs: socketPlugs,
      syncedAt: syncedAt,
    );
  }).toList(growable: false);
}
